### PR TITLE
[Merged by Bors] - Normalize keystore passwords

### LINKED ISSUE #1879: Unicode normalization
## Description

There are scenarios when managing passwords where we should be doing unicode normalization, but we're not. The same password entered on different devices can have different byte-level representations.

### Know locations:

- https://github.com/sigp/lighthouse/blob/b185d7bbd84deade495dbb5de4a57ff42d544194/common/account_utils/src/lib.rs#L106
- https://github.com/sigp/lighthouse/blob/b185d7bbd84deade495dbb5de4a57ff42d544194/account_manager/src/validator/create.rs#L255-L274
- The `rpassword` crate itself.

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

Resolves #1879 

## Proposed Changes

Do NFKD normalization for keystore passwords.

### CODE REVIEW

**Comment by pawanjay176** [comment]
@kirk-baird with https://github.com/sigp/lighthouse/pull/1928 , we are returning an error if we try encrypting using a password with control chars.
However, eip 2335 says 
> The C0, C1, and Delete control codes are not valid characters in the password and should therefore be **stripped** from the password.

This is what the deposit-cli seems to be doing here
https://github.com/ethereum/eth2.0-deposit-cli/blob/ed5a6d33a16bf7ef71a4d541212ab2896c7c0fc0/eth2deposit/key_handling/keystore.py#L116-L123

So if the user entered a password with control characters in the deposit-cli, it seems they would get stripped and fail with the import in lighthouse.

Made c5daab7 to strip the characters instead of throwing an error. Can you please check if I understood the spec correctly here? :)

**Comment by kirk-baird** [comment]
> Made c5daab7 to strip the characters instead of throwing an error. Can you please check if I understood the spec correctly here? :)

Yep I'm happy with your interpretation to strip away any control characters rather than error.

Thanks for fixing this up! I had a look through `is_control()` too and it matches exactly the control characters we wish to remove :D

**Review by kirk-baird — APPROVED** [review]
Happy with the changes :)

**Comment by pawanjay176** [comment]
Added normalization to the `eth2_keystore` instead of normalizing user input. This is in line with the eip 2335 specs.

File: crypto/eth2_keystore/src/keystore.rs [hunk 1]
```diff
@@ -331,8 +372,11 @@ pub fn encrypt(
     cipher: &Cipher,
 ) -> Result<(Vec<u8>, [u8; HASH_SIZE]), Error> {
     validate_parameters(kdf)?;
+    let mut password = normalize(password)?;
 
-    let derived_key = derive_key(&password, &kdf)?;
+    password.retain(|c| !is_control_character(c));
```
Comment (michaelsproul) [inline crypto/eth2_keystore/src/keystore.rs:377]: The comment at the top of this function

> (Errors) if `password` uses utf-8 control characters.

Is out of date

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Looks really good! Just that one doc comment to fix before merge

File: crypto/eth2_keystore/src/keystore.rs [hunk 2]
```diff
@@ -391,34 +439,20 @@ pub fn decrypt(password: &[u8], crypto: &Crypto) -> Result<PlainText, Error> {
...
-        if *char >= 0x80 && *char <= 0x9F {
-            return Err(Error::InvalidPasswordCharacter {
-                character: *char,
-                index: i,
-            });
-        }
+/// Returns true if the given char is a UTF-8 control character and false otherwise.
+fn is_control_character(c: char) -> bool {
+    // 0x00 - 0x1F + 0x80 - 0x9F + 0x7F
+    c.is_control()
```
[the remark this answers is not in the document]
Comment (michaelsproul) [reply crypto/eth2_keystore/src/keystore.rs:445]: Haha just saw Kirk's comment. Oh well, now we've triple checked it

**Review by michaelsproul — APPROVED** [review]

**Pull request closed without merging**